File: src/noiedigitalsystem/api/event/EventType.java
package noie.linmimeng.noiedigitalsystem.api.event;

/**
 * Event Type - 事件類型
 * 
 * <p>定義系統支持的事件類型。</p>
 * 
 * @since 2.0.0
 */
public enum EventType {
    /**
     * 交易事件
     * 用於資產轉移（見 NdsTransaction）
     */
    TRANSACTION,
    
    /**
     * 資產創建事件
     * 用於創建新資產
     */
    ASSET_CREATED,
    
    /**
     * 資產更新事件
     * 用於更新資產元數據
     */
    ASSET_UPDATED,
    
    /**
     * 資產刪除事件
     * 用於刪除資產
     */
    ASSET_DELETED,
    
    /**
     * 身份創建事件
     * 用於創建新身份
     */
    IDENTITY_CREATED,
    
    /**
     * 身份更新事件
     * 用於更新身份元數據
     */
    IDENTITY_UPDATED,
    
    /**
     * 系統事件
     * 用於系統級操作
     */
    SYSTEM,
    
    /**
     * 自定義事件
     * 用於擴展
     */
    CUSTOM;
    
    /**
     * 從字符串解析事件類型
     * 
     * @param str 字符串
     * @return 事件類型，如果無法解析則返回 CUSTOM
     */
    public static EventType fromString(String str) {
        if (str == null) return CUSTOM;
        try {
            return valueOf(str.toUpperCase());
        } catch (IllegalArgumentException e) {
            return CUSTOM;
        }
    }
}

File: csharp/src/Noie.Nds.Api.Abstractions/Event/EventType.cs
namespace Noie.Nds.Api.Event;

/// <summary>
/// [Index: NDS-CSHARP-EVENTTYPE-000] Event type.
/// 
/// <para>[Semantic] Standard event type taxonomy.</para>
/// </summary>
public enum EventType
{
    /// <summary>
    /// Transaction event (asset delta semantics; see INdsTransaction).
    /// </summary>
    Transaction,
    
    /// <summary>
    /// Asset created.
    /// </summary>
    AssetCreated,
    
    /// <summary>
    /// Asset updated.
    /// </summary>
    AssetUpdated,
    
    /// <summary>
    /// Asset deleted.
    /// </summary>
    AssetDeleted,
    
    /// <summary>
    /// Identity created.
    /// </summary>
    IdentityCreated,
    
    /// <summary>
    /// Identity updated.
    /// </summary>
    IdentityUpdated,
    
    /// <summary>
    /// System event.
    /// </summary>
    System,
    
    /// <summary>
    /// Custom event (extension point).
    /// </summary>
    Custom
}

/// <summary>
/// [Index: NDS-CSHARP-EVENTTYPE-100] EventType extensions.
/// </summary>
public static class EventTypeExtensions
{
    /// <summary>
    /// [Index: NDS-CSHARP-EVENTTYPE-110] Parse an event type token.
    /// </summary>
    public static EventType FromString(string? str)
    {
        if (string.IsNullOrEmpty(str)) return EventType.Custom;
        
        return str.ToUpperInvariant() switch
        {
            "TRANSACTION" => EventType.Transaction,
            "ASSET_CREATED" => EventType.AssetCreated,
            "ASSET_UPDATED" => EventType.AssetUpdated,
            "ASSET_DELETED" => EventType.AssetDeleted,
            "IDENTITY_CREATED" => EventType.IdentityCreated,
            "IDENTITY_UPDATED" => EventType.IdentityUpdated,
            "SYSTEM" => EventType.System,
            "CUSTOM" => EventType.Custom,
            _ => EventType.Custom
        };
    }
    
    /// <summary>
    /// [Index: NDS-CSHARP-EVENTTYPE-120] Convert to proto string token.
    /// </summary>
    public static string ToProtoString(this EventType type)
    {
        return type switch
        {
            EventType.Transaction => "TRANSACTION",
            EventType.AssetCreated => "ASSET_CREATED",
            EventType.AssetUpdated => "ASSET_UPDATED",
            EventType.AssetDeleted => "ASSET_DELETED",
            EventType.IdentityCreated => "IDENTITY_CREATED",
            EventType.IdentityUpdated => "IDENTITY_UPDATED",
            EventType.System => "SYSTEM",
            EventType.Custom => "CUSTOM",
            _ => "CUSTOM"
        };
    }
}
